### ISSUE TITLE
Promise.next should take Closure::<dyn FnMut(JsValue) -> JsValue>

### ISSUE THREAD

**Issue by softwareplumber** [rustwasm/wasm-bindgen#3111]
### Describe the Bug

Possible I've got the wrong end of the stick somewhere but it seems to me that Promise.next should take Closure::<dyn FnMut(JsValue) -> JsValue>

### Steps to Reproduce

Inspect signature of js_sys::Promise:

pub fn then(this: &Promise, cb: &Closure<dyn FnMut(JsValue)>) -> Promise;

### Expected Behavior

per the MDN page referenced in the source:

Once a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) is fulfilled or rejected, the respective handler function (onFulfilled or onRejected) will be called asynchronously (scheduled in the current thread loop). The behavior of the handler function follows a specific set of rules. If a handler function:

* **returns a value, the promise returned by then gets resolved with the returned value as its value.**
* **doesn't return anything, the promise returned by then gets resolved with an undefined value.**

### Actual Behavior

You can't pass in a handler that returns a value. Seems to me that a handler that returns a JsValue::UNDEFINED when it doesn't need to return anything would at least cleave a little closer to the spec behavior.

Sorry if I'm just misinterpreting something on the Rust side; these function type signatures are pretty new to me.

**Comment by linzhipeng98** [comment]
Is there any progress on this issue?

**Comment by Liamolucko** [comment]
No; the problem is that switching to the correct signature would be a breaking change, since existing code which returns `()` from the callback would then fail to compile due to the callback's new return type being `JsValue`.

The obvious (although not ideal) solution would be to add a new method with the correct signature, but I don't know what it'd be called. Maybe `then_with_result` or something?

**Comment by attilaolah** [comment]
`then_with_result` sounds right, although then you'd also have to have `then2_with_result` which is a bit awkward. Alternatively something like `transform`?

But how about introducing a breaking change and bumping the major version? Or keeping the major version but hiding the change under a feature (that way only folks with features = all would suffer).

---

Without this feature, there is really no easy way to interop with JS that requires a callback that returns a promise. Something like this is required when an Emscripten binary uses `ASYNCIFY=1` with `EM_ASYNC_JS` that then calls into a Rust wasm-bindgen binary. The Rust code here must return a Promise so it cannot itself block — except if it was somehow possible to create Promise out of an async block? I haven't figured out a way to do that though.

**Comment by daxpedda** [comment]
> But how about introducing a breaking change and bumping the major version? 

Introducing a breaking change release to `wasm-bindgen` is I believe not an option right now. We don't have enough manpower to pull this through right now.

> Or keeping the major version but hiding the change under a feature (that way only folks with features = all would suffer).

This also doesn't seem like a viable option to me, crate features should be additive and not change the API.

---

Adding a new method is the only viable way forward right now, even if it's less then ideal.

**Comment by pfernandom** [comment]
> > But how about introducing a breaking change and bumping the major version? 
> 
> Introducing a breaking change release to `wasm-bindgen` is I believe not an option right now. We don't have enough manpower to pull this through right now.
> 
> > Or keeping the major version but hiding the change under a feature (that way only folks with features = all would suffer).
> 
> This also doesn't seem like a viable option to me, crate features should be additive and not change the API.
> 
> ---
> 
> Adding a new method is the only viable way forward right now, even if it's less then ideal.

Just curious, what would you need help with for doing a major release?

**Comment by thanatos** [comment]
I just wanted to note that I'd propose `-> Result<JsValue, JsValue>` (over `-> JsValue`). A `then` handler can throw; the thrown exception causes the returned-by-`then` `Promise` to reject.

From [the MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then#return_value):
> * throws an error: `p` gets rejected with the thrown error as its value.

This makes it nicer to call other functions (e.g., JS ones) in the handler that are also `-> Result<_, JsValue>`, as you can just use `?` to propagate the error.

**Comment by guybedford** [comment]
There is a refactoring of js-sys in the new generics PR at https://github.com/wasm-bindgen/wasm-bindgen/pull/4876, which now includes this for the new functions like `then_map` and `then_with_reject`, with this original `then()` then deprecated. Further feedback on the design very welcome.
